vochain/ist/ist.go

```diff
@@ -0,0 +1,303 @@
+//
+// IST actions are encoded and decoded using the Gob encoding package.
+// The encoding and decoding functions are provided by the Actions type.
+//
+// The Controller struct provides methods for scheduling, rescheduling, and committing
+// IST actions. When committing an action, it performs the necessary operations depending
+// on the type of action, such as computing results, committing results, or ending processes.
+// It also provides methods to store and retrieve data from the no-state, which is a separate
+// key-value store that holds non-consensus-critical data.
+//
+// Constants used in the package are:
+//
+// - BlocksToWaitForResultsFactor: a factor used to compute the height at which the results will be committed to the state.
+// - ExtraWaitSecondsForResults: the number of extra seconds to wait for results.
+//
+// Technical decisions taken in the code include:
+//
+// - Implementing a separate no-state key-value store for non-consensus-critical data storage.
+// - Computing results in a separate goroutine to avoid blocking state transitions.
+// - Including a mechanism for rescheduling actions in case of missing keys or other issues.
+package ist
+
+import (
+	"bytes"
+	"encoding/gob"
+	"encoding/hex"
+	"errors"
+	"fmt"
+
+	"go.vocdoni.io/dvote/db"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/vochain/results"
+	"go.vocdoni.io/dvote/vochain/state"
+	"go.vocdoni.io/proto/build/go/models"
+)
+
+const (
+	dbPrefix = "ist/"
+)
+
+// Controller is the internal state transition controller.
+type Controller struct {
+	st              *state.State
+	isSynchronizing uint32
+}
+
+// NewISTC creates a new ISTC.
+// The ISTC is the controller of the internal state transitions.
+// It schedule actions to be executed at a specific height that will modify the state.
+func NewISTC(s *state.State) *Controller {
+	return &Controller{
+		st: s,
+	}
+}
+
+const (
+	// ActionComputeResults computes and schedules the commit of results, for the given election and height.
+	ActionComputeResults = iota
+	// ActionCommitResults schedules the commit of the results to the state.
+	ActionCommitResults
+	// ActionEndProcess sets a process as ended. It schedules ActionComputeResults.
+	ActionEndProcess
+)
+
+// ActionsToString translates the action identifers to its corresponding human friendly string.
+var ActionsToString = map[int]string{
+	ActionComputeResults: "compute-results",
+	ActionCommitResults:  "commit-results",
+	ActionEndProcess:     "end-process",
+}
+
+// Actions is the model used to store the list of IST actions for
+// a specific height into state.
+type Actions struct {
+	Actions map[string]Action
+}
+
+// Action is the model used to store the IST actions into state.
+type Action struct {
+	Action     int
+	ElectionID []byte
+	Attempts   uint32
+}
+
+// encode performs the encoding of the IST action using Gob.
+func (m *Actions) encode() []byte {
+	var buf bytes.Buffer
+	if err := gob.NewEncoder(&buf).Encode(m); err != nil {
```

Comment:
remember that encoding/gob is not a stable format; it may change with future Go versions. I think it's fine for in-memory stuff or temporary caches, but anything relatively permanent should use a stable format. For that, I'd either use go-bare if we don't need a schema, or protobuf if we want a schema.

Reply:
Since this information is not relevant for the Consensus State (its stored into a NoState() KV space), it is not a big issue if the Gob encoder in some point changes to a different format.